### DISCUSSION TITLE
Understanding appindicator_tray.py

### DISCUSSION THREAD

**Discussion by sbradnick** [Xpra-org/xpra#4181] [Q&A] [answered]
I'm working with eww and the newly added systray functionality. Keeping my expectations in check considering i3+eww (x11) probably isn't high on the support list and without getting too far into the weeds out of the gate, running `python3 /usr/lib64/python3.11/site-packages/xpra/platform/posix/appindicator_tray.py` appears completely functional. Is there a way to force that to be used? I'm not sure if the code is using `GTKStatusIconTray` no-matter-what at this point.

The eww log shows this entry when I run the `appindicator_tray.py` command: `notifier_host::watcher    > new item: :1.269/org/ayatana/NotificationItem/test`

Thanks!

**Comment by totaam** [comment]
Yes, this needs to be explained, because choosing a backend is just really nasty.

First, there are two types of system trays.
## Xpra's own menu
Configured using (`--tray=yes|no`)
It is handled in https://github.com/Xpra-org/xpra/blob/master/xpra/client/mixins/tray.py
The actual tray is instantiated here:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/mixins/tray.py#L122-L123
Using the tray backends from:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/mixins/tray.py#L80-L84
This method is overriden in the GTK client implementation to add `Gtk.StatusIcon` to the backend list:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/gtk3/client_base.py#L818-L820
Whether it goes first or last in the backend list of options is decided here:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/gtk3/client_base.py#L805-L816
We don't use it all on Wayland, and it goes last in the list with gnome, macos and MS Windows.

## system tray forwarding
Configured using `--system-tray=yes|no`
This is handled by the generic window forwarding code: https://github.com/Xpra-org/xpra/blob/master/xpra/client/mixins/windows.py
It follows a similar path:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/mixins/windows.py#L627-L631
Getting its backends from:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/mixins/windows.py#L634-L638
Again, with the GTK client adding `Gtk.StatusIcon` as before:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/client/gtk3/client_base.py#L822-L820

It used to be more complicated. (Unity misbehaving, Ubuntu and Debian breaking random things)
And I'm still not happy with this ridiculous dance we have to go through:
https://github.com/Xpra-org/xpra/blob/08ce2221bf560530e42fc70fe1a6a11f6e53c262/xpra/platform/posix/appindicator_tray.py#L27-L37

So, I guess we should at least add an environment variable to allow users to choose the backend more easily.
And perhaps the window manager detection could be improved using an allow / deny list for each.
`xpra wminfo` does give us data we can use for the decision making.

**Comment by totaam** [comment, marked as answer]
With 7573ac099f3c35da2dc2af63f266810366d62b8f, now you can do:
```
XPRA_PREFER_STATUSICON=0 xpra attach ...
```

**Comment by sbradnick** [comment]
Works great! :)

Here's the output from that command, if it's of any use.
```
$ xpra wminfo
root=1362
WM_S0=3145730
_NET_SUPPORTING_WM_CHECK=3145741
wmname=i3
display=:0
```

**Reply by totaam** [reply]
Second thought, `wminfo` requires shelling out to be safe.
How about, `env | grep XDG` ?

**Reply by sbradnick** [reply]
Here's the output from the client, which is now making use of using eww's systray mechanism :+1: 
```
$ env | grep XDG
XDG_SESSION_PATH=/org/freedesktop/DisplayManager/Session0
XDG_SESSION_CLASS=user
XDG_VTNR=7
XDG_SESSION_TYPE=x11
XDG_SEAT=seat0
XDG_RUNTIME_DIR=/run/user/1000
XDG_SESSION_ID=3
XDG_SEAT_PATH=/org/freedesktop/DisplayManager/Seat0
XDG_CURRENT_DESKTOP=i3
XDG_GREETER_DATA_DIR=/var/lib/lightdm-data/scott
XDG_SESSION_DESKTOP=i3

```
And output from an xpra connection's shell:
```
$ env | grep XDG
XDG_CURRENT_DESKTOP=Xpra
XDG_RUNTIME_DIR=/run/user/1000
XDG_CONFIG_DIRS=/etc/xdg:/usr/local/etc/xdg:/usr/etc/xdg
XDG_SESSION_TYPE=x11
```

**Reply by totaam** [reply]
FWIW, I've changed the default in v6: 0ffeab68dc76f9e37d8b4c3caf3fe0d47803951e
It seemed safer than an always incomplete list of window managers.
It's a shame because `Gtk.StatusIcon` is much more usable for tray forwarding that any of the "newer" (supposedly "better") alternatives.

**Reply by sbradnick** [reply]
As of using e2d97e37 I can omit `env XPRA_PREFER_STATUSICON=0` now when I run `xpra attach` and I still get the xpra tray icon so I can interact with the remote session :+1:
